TFNetworkLayer.py

```diff
@@ -4750,21 +4750,23 @@ def __init__(self, mode, axes=None, axis=None, keep_dims=False, enforce_batch_di
           mask = tf.logical_and(mask, tf.ones_like(x_, dtype=mask.dtype))
 
           replacement_value = {
+            tf.reduce_mean: tf.zeros_like(x.placeholder),
             tf.reduce_sum: tf.zeros_like(x.placeholder),
             tf.reduce_min: tf.zeros_like(x.placeholder) + x.placeholder.dtype.max,
             tf.reduce_max: tf.zeros_like(x.placeholder) + x.placeholder.dtype.min
           }
 
           x_ = tf.where(mask, x_, replacement_value[f], "x_masked_axis_%i" % axis)
-      elif f == tf.reduce_mean:
-        # Flattening.
-        if x.time_dim_axis in axes:
-          assert not keep_dims, "not yet implemented otherwise"
-          assert x.batch_dim_axis in axes, "not yet implemented otherwise"
-          axes = [a if (a < x.time_dim_axis) else (a - 1)
-                  for a in axes if a != x.time_dim_axis]
-          x = x.copy_time_flattened()
-          x_ = x.placeholder
+          if f == tf.reduce_mean:
+            seq_len_bc = tf.reshape(
+              x.get_sequence_lengths(),
+              [1 if (i != x.batch_dim_axis) else -1 for i in range(x.batch_ndim)])  # (1,..B..,1)
+            x_ = x_ / tf.cast(seq_len_bc, tf.float32)
```

Comment:
This should not be done multiple times (if this code is run for multiple axes). And this only really make sense if `axis == x.time_dim_axis`.
I think I really would write specific code for `axes == [x.time_dim_axis]`. Otherwise this is a bit too unreadable.

Reply:
It's still not so nice. If you read only this code, and not the check a couple of lines above (`axes == [x.time_dim_axis]`), you would think this is a bug. Code should be locally bug-free, as far as possible, e.g. when you remove the check `axes == [x.time_dim_axis]` above, this should still be correct. (This is for easier readability, but also for maintainability, e.g. when changes to the code are made, e.g. someone might think, why do we have this check `axes == [x.time_dim_axis]` there, it should also work without, and then removes it...)
Anyway, I only see the diff here, and don't really know whether other ways to write this are more clear, so let's leave it like that for now.